File: dramatiq/drain.py
import os
import signal
import sys
import time

from redis import Redis

num_queues = int(os.getenv("NUM_QUEUES", 10))
poll_interval = int(os.getenv("POLL_INTERVAL", 2))
worker_pid = int(sys.argv[1])

r = Redis()
queue_keys = [f"{{dr}}:queue_{i}" for i in range(num_queues)]


def drain_and_shutdown():
    while True:
        total = sum(r.llen(k) for k in queue_keys)
        if total == 0:
            print("All queues empty. Shutting down...")
            os.kill(worker_pid, signal.SIGTERM)
            break
        time.sleep(poll_interval)


if __name__ == "__main__":
    drain_and_shutdown()

File: sidekiq/drain.rb
require "redis"

poll_interval = Integer(ENV.fetch("POLL_INTERVAL", 2))
num_queues = Integer(ENV.fetch("NUM_QUEUES", 10))
worker_pid = Integer(ARGV[0])

redis = Redis.new

queue_keys = (0...num_queues).map { |i| "queue:queue_#{i}" }

loop do
  total = queue_keys.sum { |k| redis.llen(k) }
  if total == 0
    puts "All queues empty. Shutting down..."
    Process.kill("TERM", worker_pid)
    break
  end
  sleep poll_interval
end
